src/agentlab/llm/llm_utils.py

```diff
@@ -322,6 +323,149 @@ def image_to_jpg_base64_url(image: np.ndarray | Image.Image):
+        if not all(elem["type"] == "text" for elem in self["content"]):
+            logging.warning(
+                "The content of the message has images, which are not displayed in the string representation."
+            )
+        return "\n".join([elem["text"] for elem in self["content"] if elem["type"] == "text"])
+
+    def add_content(self, type: str, content: Any):
+        if isinstance(self["content"], str):
+            text = self["content"]
+            self["content"] = []
+            self["content"].append({"type": "text", "text": text})
+        self["content"].append({"type": type, type: content})
+
+    def add_text(self, text: str):
+        self.add_content("text", text)
+
+    def add_image(self, image: np.ndarray | Image.Image | str, detail: str = None):
+        if not isinstance(image, str):
+            image_url = image_to_jpg_base64_url(image)
+        else:
+            image_url = image
+        if detail:
+            self.add_content("image", {"url": image_url, "detail": detail})
+        else:
+            self.add_content("image", image_url)
+
+    def to_markdown(self):
+        if isinstance(self["content"], str):
+            return f"\n```\n{self['content']}\n```\n"
+        res = []
+        for elem in self["content"]:
+            # add texts between ticks and images
+            if elem["type"] == "text":
+                res.append(f"\n```\n{elem['text']}\n```\n")
+            elif elem["type"] == "image":
+                res.append(f"![image]({elem['url']})")
+        return "\n".join(res)
+
+
+class SystemMessage(BaseMessage):
+    def __init__(self, content: Union[str, list[dict]]):
+        super().__init__("system", content)
+
+
+class HumanMessage(BaseMessage):
+    def __init__(self, content: Union[str, list[dict]]):
+        super().__init__("user", content)
+
+
+class AIMessage(BaseMessage):
+    def __init__(self, content: Union[str, list[dict]]):
+        super().__init__("assistant", content)
+
+
+class Discussion:
+    def __init__(self, messages: Union[list[BaseMessage], BaseMessage] = []):
+        if isinstance(messages, BaseMessage):
+            messages = [messages]
+        self.messages = messages
+
+    @property
+    def last_message(self):
+        return self.messages[-1]
+
+    def __str__(self) -> str:
+        return "\n".join(str(m) for m in self.messages)
+
+    def to_string(self):
+        return str(self)
+
+    def to_openai(self):
+        return self.messages
+
+    def add_message(
+        self,
+        message: BaseMessage | dict = None,
+        role: str = None,
+        content: Union[str, list[dict]] = None,
+    ):
+        if message is None:
+            message = BaseMessage(role, content)
+        else:
+            if isinstance(message, dict):
+                message = BaseMessage(**message)
+        self.messages.append(message)
+
+    def append(self, message: BaseMessage | dict):
+        self.add_message(message)
```

Comment:
This makes things even more retro-compatible, leaving many previous functions untouched